homeassistant/components/crownstone/hub.py

```diff
@@ -0,0 +1,157 @@
+from crownstone_sse import CrownstoneSSE
+from crownstone_sse.const import (
+    EVENT_ABILITY_CHANGE_DIMMING,
+    EVENT_ABILITY_CHANGE_SWITCHCRAFT,
+    EVENT_ABILITY_CHANGE_TAP_TO_TOGGLE,
+)
+from crownstone_sse.events.AbilityChangeEvent import AbilityChangeEvent
+
+from homeassistant.config_entries import ConfigEntry, ConfigEntryNotReady
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.helpers import aiohttp_client
+
+from .const import CONF_SPHERE, LIGHT_PLATFORM
+from .helpers import UartManager
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class CrownstoneHub:
+    """Manage all Crownstone IO."""
+
+    def __init__(self, hass, config_entry) -> None:
+        """Initialize the hub."""
+        self.sphere: Optional[Sphere] = None
+        self.config_entry: ConfigEntry = config_entry
+        self.hass: HomeAssistant = hass
+        self.cloud: Optional[CrownstoneCloud] = None
+        self.uart_manager: Optional[UartManager] = None
+        self.sse: Optional[CrownstoneSSE] = None
+
+    async def async_setup(self) -> bool:
+        """
+        Set up the Crownstone hub.
+
+        The hub is a combination of Crownstone cloud, Crownstone SSE and Crownstone uart.
+        Returns True if the setup was successful.
+        """
+        # Setup email and password gained from config flow
+        customer_email = self.config_entry.data[CONF_EMAIL]
+        customer_password = self.config_entry.data[CONF_PASSWORD]
+
+        # Create cloud instance
+        self.cloud = CrownstoneCloud(
+            email=customer_email,
+            password=customer_password,
+            websession=aiohttp_client.async_get_clientsession(self.hass),
+        )
+        # Login
+        try:
+            await self.cloud.async_initialize()
+        except CrownstoneAuthenticationError as auth_err:
+            _LOGGER.error(
+                "Auth error during login with type: %s and message: %s",
+                auth_err.type,
+                auth_err.message,
+            )
+            return False
+        except CrownstoneUnknownError as unknown_err:
+            _LOGGER.error("Unknown error during login")
+            raise ConfigEntryNotReady from unknown_err
+
+        # set the sphere we chose to setup in the flow
+        self.sphere = self.cloud.spheres.find(self.config_entry.data[CONF_SPHERE])
+
+        # Create uart manager to manage usb connections
+        # uart.is_ready() returns whether the usb is ready or not.
+        self.uart_manager = UartManager()
+        self.uart_manager.start()
+
+        # Create SSE instance
+        self.sse = CrownstoneSSE(customer_email, customer_password)
+        self.sse.set_access_token(self.cloud.get_access_token())
+        self.sse.start()
+
+        # subscribe to Crownstone ability updates
+        self.sse.add_event_listener(EVENT_ABILITY_CHANGE_DIMMING, self.update_ability)
```

Comment:
Is SSE calling event listeners from the event loop or a thread?

Reply:
They are called from a thread.

Reply:
Then don't decorate with `@callback` and don't call any `async` function.

Reply:
is it safe to initialize the SSE thread with the mainthread loop like so
```python
self.sse = CrownstoneSSE(
      customer_email, customer_password, asyncio.get_running_loop()
)
```
And then in the eventbus, provide coroutines to the loop by calling
```python
asyncio.run_coroutine_threadsafe(coroutine, loop)
```

Reply:
If it is a thread, why do you have to pass in the loop?

Reply:
The SSE lib has it's own eventbus. Which means any callbacks or coroutines added to it are executed within that thread (and the event loop running there). Let's say i want to run
```python
await hass.async_create_task(coroutine)
```
in a listener in that thread. Do i need to pass it to the mainthread loop specifically or am I not understanding it correctly?